### PR TITLE
Liisa NFT Portfolio Tracker

### PR DESCRIPTION
**Pull request by LiisaNFT**
#### Project Abstract

This project consists of the development of a custom-built portfolio tracker, explicitly devised for NFTs within the Polkadot blockchain network, catering to the specificities of NFT investment strategies.

Within the Substrate/Kusama/Polkadot/Web 3 ecosystem, Liisa intends to emerge as a central supplier of granular NFT analytics, distinguishing itself with a user interface tailored to meet the specific requirements of NFT investors.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application, and sorry for the late reply here. I have a few initial questions/comments:
- Could you add the programming language to the specification of each deliverable? This helps us to understand what exactly will be developed. 
- One concern I have is that as part of the grant, you mostly rely on subscan. This basically means we would support subscan with this project rather than being entirely objective. 
- Another concern is that all the deliveries are highly specific to your project, which so far seems to be closed-source. 
- Finally, we usually try to focus on smart contract projects that are native to our ecosystem rather than EVM-based solutions. 

Could you comment on the above? And feel free to update the application accordingly.

**Comment by LiisaNFT** [comment]
> Thanks for the application, and sorry for the late reply here. I have a few initial questions/comments:
> 
> * Could you add the programming language to the specification of each deliverable? This helps us to understand what exactly will be developed.
> * One concern I have is that as part of the grant, you mostly rely on subscan. This basically means we would support subscan with this project rather than being entirely objective.
> * Another concern is that all the deliveries are highly specific to your project, which so far seems to be closed-source.
> * Finally, we usually try to focus on smart contract projects that are native to our ecosystem rather than EVM-based solutions.
> 
> Could you comment on the above? And feel free to update the application accordingly.

Thank you for your feedback David. Here are our comments to your 4 points:

1) We've incorporated the specification of programming languages into our deliverable requirements

2) Your concern regarding our choice of Subscan for the MVP is completely valid. Our decision is primarily rooted in the interest of expediency. Leveraging an established third-party solution like Subscan, which already has a robust infrastructure for efficient querying, structuring, and indexing on-chain NFT data, enables us to realize a working MVP more swiftly. However, we remain open to alternatives that can deliver comparable speed without necessitating third-party involvement.

3) Initially, our platform was developed as a closed-source system. Nevertheless, our ultimate goal is to transition most components to open source. We've already started this process by migrating a substantial portion of our backend code to Github. Please note that we anticipate that the backend code and Bubble.io plugins we're developing for the portfolio tracker  (which will be open-sourced and made available for free to all developers in the ecosystem) will have a broader utility beyond our project, proving beneficial to other entities such as analytical tools, marketplaces, and wallet providers. As the ecosystem matures, we foresee standard industry metrics for NFT trading, like rarity, price and volume fluctuations, becoming the norm.

4) We consulted with the Parity Data team before initiating our application and gained some valuable insights. Given our team's prior experience with EVM chains, it was collectively agreed that we should concentrate exclusively on these for the initial MVP. This strategic approach not only accelerates the delivery of a functional MVP but also lays a strong foundation for future iterations that will incorporate other para-chains. Please note, our tool's primary function is data retrieval from the chains; we do not engage in smart contract development.

Kindly let us know if we should make any of these more explicit in the body of the application.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the very quick and detailed reply here. Could you potentially clarify or add as part of milestone two deliveries that the backend will be open-sourced? Regarding the subscan deliveries, maybe you can try to make it modular in a way that makes it easy to replace the API integration with another one and also provide documentation regarding this and add it to the deliveries of milestone 1.

**Comment by LiisaNFT** [comment]
> Thanks for the very quick and detailed reply here. Could you potentially clarify or add as part of milestone two deliveries that the backend will be open-sourced? Regarding the subscan deliveries, maybe you can try to make it modular in a way that makes it easy to replace the API integration with another one and also provide documentation regarding this and add it to the deliveries of milestone 1.

Dear David, thank you very much for your feedback. 

We have changed the Components section on item 1) to express that we will develop the API integration in a modular way in order for it to be easy to replace it with another one. Regarding the documentation for this, it would be included as part of deliverable "0b. Documentation" from Milestone 1. 

Regarding open sourcing the backend code, we have made it clear in the Components section, item 2), third paragraph, that all this code will be open-sourced. Furthermore, we have also selected Apache 2.0 license for both Milestone 1 and Milestone 2.

Do you see any further changes to be made?

Looking forward to hearing from you,
Liisa Team

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. Could you also update/change the milestone tables? According to our terms and conditions, the milestone tables list the requirements of our contracts. Evaluators also will primarily focus on these deliverables during the evaluation.

File: applications/LiisaPortfolioTracker.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+#### 4) Frontend
+
+Within the context of the MVP, we'll utilize bubble.io for frontend development to accelerate development speed and iterate quickly. This approach will help us create a functional product that meets user needs faster. Upon achieving a working product, at a later stage we will transition to react.js for enhanced flexibility and performance.
+
+#### 5) Bubble.io plug-ins
+
+We will develop open-source plug-ins for Bubble.io, enabling connections to the selected wallets: Subwallet, Parity signer, Polkadot-js, Subsquid, Fearless wallet, Nova Wallet, Talisman. Built using Node.js, these plug-ins will allow for the users to log-in to the platform using their wallets, and will facilitate the retrieval of information about the NFTs owned by the user to compute the portfolio metrics. By open-sourcing these plug-ins, we intend to improve access from the growing no-code community to the Polkadot ecosystem.
```
Comment (takahser) [inline applications/LiisaPortfolioTracker.md:68]: How would your polkadot js plugin differ from the existing: https://bubble.io/plugin/polkadot-wallet-1639402639641x977692461648052200

File: applications/LiisaPortfolioTracker.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+![Portfolio tracker #1 (1)](https://github.com/LiisaNFT/W3F-Grants-Program/assets/139144686/ae122745-4d96-4c2a-86af-a8bc8230f63a)
+
+### NFTs page
+
+The NFTs page offers users a technical and visually engaging interface to view their NFT collection:
+- NFT Images: Displaying artwork and associated metadata for each token;
+- Token Metrics: Showing estimated value, acquisition price, unrealized PNL, and 24-hour sales for each collection;
+- Filtering and Sorting: Allowing users to sort and filter NFTs based on metrics like collection sales and estimated value.
+
+![Portfolio tracker #1 (1)](https://github.com/LiisaNFT/W3F-Grants-Program/assets/139144686/c8b1dde7-d548-4153-92d7-064b4f0f3ada)
```
Comment (takahser) [inline applications/LiisaPortfolioTracker.md:218]: I think you posted this image already on [L209](https://github.com/w3f/Grants-Program/pull/1855/files#diff-7b9b655bfcc02275c2dc8f9b9bd14841955ca499c300867de9f29ecdb1437b04R209)

**Review by takahser — CHANGES_REQUESTED** [review]
I have concerns about the frontend being implemented in bubble.io which is a proprioritary no-code platform. Quote from our [guidelines](https://github.com/w3f/Grants-Program/#guidelines):

> All code produced as part of a grant must be open-sourced, and it must also not rely on closed-source software for full functionality.

Regarding the NFTs you're going to support, are you focusing solely on EVM-based NFTs or are you going to support any other NFTs that are native to our ecosystem? For reference here is a list of ecosystem-native NFT projects: https://wiki.polkadot.network/docs/learn-nft-projects

Finally, feel free to have a look at my inline comments.

**Comment by LiisaNFT** [comment]
Thanks for your valuable feedback @takahser . Please find below our answers to your questions.

**_Why have we chosen bubble.io for the frontend?_**

Our choice for bubble.io was motivated by the opportunity for rapid iteration in response to early user feedback and seamless integration with our existing platform. Furthermore, this would've opened doors for the Polkadot ecosystem by introducing new Polkadot wallet plugins on the bubble.io platform.

However, we fully understand and respect the guidelines. To meet them, we can transition our frontend development to React.js. While this approach guarantees a straightforward open-source process, it does come at the cost of a slightly extended development timeline. More significantly, we would be foregoing the development of the Bubble plugins — a potential game-changer in expanding the Polkadot ecosystem and enriching user experience. While not creating these plugins does free up time to expedite the React.js development, we believe the unique advantages offered by the bubble.io integration could have profound impacts on the project's success and user adoption.

We're keen on aligning with the best interests of both the project and the grant's conditions. Please advise on the preferred approach.

**_Existing polkadot.js wallet plugin for bubble.io_**

We were not previously aware of the existence of this plug in. We believe that if this plugin already provides the desired functionality for Polkadot.js wallet, it would be more practical to utilize it instead of developing a new one from scratch. The plugin seems to align perfectly with our initial idea. However, we plan to extend this concept to encompass support for the other wallets mentioned.

**_Duplicated image_**

.md file was updated with the correct image for the NFTs page.

**Comment by LiisaNFT** [comment]
> I have concerns about the frontend being implemented in bubble.io which is a proprioritary no-code platform. Quote from our [guidelines](https://github.com/w3f/Grants-Program/#guidelines):
> 
> > All code produced as part of a grant must be open-sourced, and it must also not rely on closed-source software for full functionality.
> 
> Regarding the NFTs you're going to support, are you focusing solely on EVM-based NFTs or are you going to support any other NFTs that are native to our ecosystem? For reference here is a list of ecosystem-native NFT projects: https://wiki.polkadot.network/docs/learn-nft-projects
> 
> Finally, feel free to have a look at my inline comments.

Regarding your inquiry about the NFTs we intend to support:

In our prior discussions with the Parity team, it was emphasized that scoping the project was crucial to adhere to time and budget constraints for this first phase. This necessitated selecting specific parachains to focus on. The overarching goal is to first launch a robust product, refine the user experience, and subsequently broaden our support. We were guided towards prioritizing EVM-compatible parachains initially, given our team's existing proficiency with EVM NFTs.

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application, @LiisaNFT. To add to what David and Seraya already said, I have a few questions as well:

- Do you have any experience writing and maintaining open source software? I checked the Github profiles and didn't find any activity on any in the past few years. 
- The comparison of your project to others is very high-level. As you know, this program is for technical projects, so it'd be great if you could go into a bit more detail about what functionality or part of the architecture might be new.

**Comment by LiisaNFT** [comment]
> Thanks for the application, @LiisaNFT. To add to what David and Seraya already said, I have a few questions as well:
> 
> * Do you have any experience writing and maintaining open source software? I checked the Github profiles and didn't find any activity on any in the past few years.
> * The comparison of your project to others is very high-level. As you know, this program is for technical projects, so it'd be great if you could go into a bit more detail about what functionality or part of the architecture might be new.

Dear Sebastian,

Thank you for taking the time for reviewing the application. Here are our clarifications for your questions: 

1) Open source experience

Our team is well-acquainted with GitHub, ensuring proficiency in version control, pull requests, and collaborative repository management. Kerim, one of our team members, has hands-on open-source experience from his previous work in Turkey. The rest of the team is still familiarizing itself with the open-source community.

We have historically hosted our code on private AWS instances, which limited visibility into our past activity. We are now in the process of transitioning parts of our codebase to public GitHub repositories to promote greater transparency and foster collaboration. 

To ensure we align with the standards of your program, we are available to do the following:
- Engaging with experienced open-source developers/teams for guidance;
- Actively seeking and incorporating your feedback to refine our approach;
- Provide the team with training on opensource methods and best practices.

2) Differentiation

Thank you for your feedback. We acknowledge our oversight in the preliminary comparison, and after a more detailed assessment, here's our differentiation:

Contrary to our initial analysis, Nansen and Coinstats do not support Polkadot NFTs (they only support tokens), and De.Fi's status remains uncertain. We found a conspicuous gap in the Polkadot ecosystem—there isn't a dedicated NFT portfolio tracker. Our project aims to fill this void.

Since we could not find direct competition, we can compare ourselves in regards to technical differentiation to wallet services.

Technical Differentiation:
- Users get a comprehensive summary including a unique Portfolio Scorecard assessing liquidity, volatility, and NFT diversification;
- Beyond visual displays, we provide dynamic filtering and sorting based on sales and estimated value metrics, offering both a technical and engaging interface;
- Detailed insights for each NFT token, from rarity rank within a collection to an activity table logging all related events;
- Aggregated view of NFT holdings per collection, integrating quick search and dynamic filtering, offering a novel way for users to understand collection performance.

In a nutshell, our project introduces new metrics, visualization techniques, and insights into the NFT space, previously unavailable in the Polkadot ecosystem.

**We hope these answer your questions. Kindly let us know if this is clear and if we should update the application accordingly**

**Comment by nikw3f** [comment]
Instead of subscan i would recommend to use https://www.subsquid.io/. It is open source and built for general data ETL pipeline for dapps. Its also free at the moment.

**Review by takahser — CHANGES_REQUESTED** [review]
@LiisaNFT while I understand your choice for implementing the frontend on bubble.io rather than coding it from a business perspective my guess is the community won't be able to reuse any of it. As already mentioned, our program is tailored for open source efforts, hence I won't support a bubble-based implementation.

Regarding the development of the wallet plugins for bubble.io, I do believe that it could be worth considering it even though the [existing plugin](https://bubble.io/plugin/polkadot-wallet-1639402639641x977692461648052200) hasn't seen much traction yet:

image

Besides that, I think it'd be worth taking into consideration @nikw3f's [comment](https://github.com/w3f/Grants-Program/pull/1855#issuecomment-1680600230) regarding replacing Subscan with Subsquid. Keep in mind that any milestone delivery must not rely on any closed-source software and it seems like Subscan is not fully open-source.

Lastly, I do share the concerns expressed by @semuelle regarding your team's experience in open-source development. Since the github profiles show no activity I think it'd be good to provide us with some references of what your team has built in the past, just to get a better understanding of your capabilities.

**Review by takahser — CHANGES_REQUESTED** [review]
@LiisaNFT in reply to your comment:

> The team has discussed and we can use React.js instead of Bubble.io for the frontend.

That's great, thanks for your flexibility here!

> As per your feedback, we'll still develop the Bubble plugins, as we believe these are beneficial to the future of the ecosystem, and will be required for future integration of Polkadot wallets in our existing platform.

Yes, that might be useful. Although it might sound a bit confusing since these are essentially two different projects, I'd be open to support that decision, given the potential usefulness to bubble users.

> We are happy to replace Subscan with Subsquid. Please bear in mind that, in that case, the protocols that we will be supporting within the scope of this project will be Astar, Moonbeam EVM, Moonriver EVM, Moonsama EVM. These are the protocols supported by Subsquid.

I'm not sure if that's correct - their [examples list](https://docs.subsquid.io/firesquid/examples/substrate/list/) includes a [Simple ink!/WASM squid](https://github.com/subsquid-labs/squid-wasm-template) and there is also a [ink! contract indexing
tutorial](https://docs.subsquid.io/tutorials/create-a-wasm-processing-squid/).

**Comment by LiisaNFT** [comment]
recheck

**Comment by LiisaNFT** [comment]
I have read and hereby sign the Contributor License Agreement.

**Comment by LiisaNFT** [comment]
recheck

File: applications/LiisaPortfolioTracker.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+- **FTE:**  2.5
+- **Costs:** 17,000 USDC
+
+| Number | Deliverable               | Specification |
+| ------ | ------------------------- | ------------- |
+| 0a.    | License                   | All code will be open-sourced under Apache 2.0 license. |
+| 0b.    | Documentation             | Documentation includes Inline Code Documentation, Configuration Documentation, Readme file. Documentation on the modular design of Subsquid API calls is included. |
+| 0c.    | Testing Guide             | The code will have unit-test coverage (min. 50%) to ensure functionality and robustness. In the guide, we will describe how to run these tests |
+| 0d.    | Docker                    | Provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1a.    | Modular Subsquid API calls | Design and implement API queries to extract event data from the Subsquid API for the supported protocols: Astar, Basilisk, Efinity, Moonbeam, Unique Network, Moonriver  and Moonsama. The design will be modular to allow easy replacement with different APIs. This will be developed using Node.js. |
```
Comment (takahser) [inline applications/LiisaPortfolioTracker.md:182]: > Design and implement API queries to extract event data from the Subsquid API for the supported protocols: Astar, Basilisk, Efinity, Moonbeam, Unique Network, Moonriver and Moonsama. The design will be modular to allow easy replacement with different APIs. This will be developed using Node.js.

Are you going to support all NFTs on these networks? No matter what standards or pallet's they'd use? Please clarify directly within the Spec here.

**Review by takahser — CHANGES_REQUESTED** [review]
@LiisaNFT thanks, it looks much better now. I have just one last thing (see my inline comment}.

File: applications/LiisaPortfolioTracker.md (hunk 3 above)
Comment (LiisaNFT) [reply applications/LiisaPortfolioTracker.md]: Dear @semuelle ,

Thank you for your insightful feedback. We agree that scoping the project clearly is critical for its success. To address your question about supporting all NFTs on the networks we listed, we have made specific choices in line with our platform's capabilities and the project budget.

For our initial MVP, we will support the following NFT standards for each protocol:

Astar Network: ERC-721, ERC-1155, PSP 34 (similar to ERC-721), and PSP 37 (similar to ERC-1155)

Basilisk: We will support the Uniques pallet, which is straightforward and standardized.

Efinity: ERC-721 and ERC-1155. While Efinity does support additional standards, we will initially focus on these two.

Moonbeam: ERC-721 and ERC-1155.

Unique Network: We'll support their native NFTs.

Moonriver: ERC-721 and ERC-1155.

Moonsama: ERC-721 and ERC-1155.

The reason for these choices is mainly that our platform is already compatible with ERC-721 and ERC-1155 standards. The other pallets and standards we selected are either identical or similar enough that we can integrate them easily.

We hope this clarifies our approach for the MVP. As the project evolves, we are certainly open to expanding the range of supported standards based on community demand and technical feasibility.

Thank you once again for your thoughtful questions; they help make our project better.

Best regards,
Liisa Team

Comment (takahser) [reply applications/LiisaPortfolioTracker.md]: Lovely, thanks for the update.

**Review by takahser — APPROVED** [review]
@LiisaNFT thanks for the work you've put into this. At this stage I'm happy to approve it.

**Review by Noc2 — APPROVED** [review]
